replica/out_of_space_controller.cc

```diff
@@ -0,0 +1,74 @@
+/*
+ * Copyright (C) 2025-present ScyllaDB
+ */
+
+/*
+ * SPDX-License-Identifier: LicenseRef-ScyllaDB-Source-Available-1.0
+ */
+#include "replica/out_of_space_controller.hh"
+#include "utils/log.hh"
+
+namespace replica {
+
+static logging::logger logger("out_of_space_controller");
+
+out_of_space_controller::out_of_space_controller(utils::disk_space_monitor& dsm, config cfg, abort_source& as)
+    : _cfg(std::move(cfg))
+    , _abort_source(as)
+    , _dsm(dsm)
+    , _dsm_subscription(dsm.listen([this](const utils::disk_space_monitor& dsm) -> future<> {
+        if (_abort_source.abort_requested()) {
+            co_return;
+        }
+
+        const float current_disk_utilization = dsm.disk_utilization();
+        if (current_disk_utilization < 0.0f) {
+            co_return;
+        }
+
+        logger.debug("current disk utilization={}", current_disk_utilization);
+
+        const bool old = _critical_disk_utilization_threshold_reached;
+        _critical_disk_utilization_threshold_reached = current_disk_utilization > std::clamp(_cfg.critical_disk_utilization_threshold(), 0.0f, 1.0f);
+
+        if (old == _critical_disk_utilization_threshold_reached) {
+            co_return co_await notify_only_newly_registered();
+        }
+
+        static constexpr auto msg_template = "{} the critical disk utilization level ({:.1f}%%). Current disk utilization {:.1f}%%";
+        if (_critical_disk_utilization_threshold_reached) {
+            logger.warn(msg_template, "Reached", _cfg.critical_disk_utilization_threshold() * 100, current_disk_utilization * 100);
+        } else {
+            logger.info(msg_template, "Dropped below", _cfg.critical_disk_utilization_threshold() * 100, current_disk_utilization * 100);
+        }
+
+        co_await notify();
+    }))
+    {
+        dsm.trigger_poll();
```

Comment:
Why is this needed if there are no subscribers yet?

Reply:
I initially needed it for the unit tests but you are right. Most likely it's a left over after the re-write of the controller. I will double check and remove it.